top_hits.py

```diff
@@ -0,0 +1,59 @@
+import argparse
+import os
+
+import numpy as np
+
+
+def find_row_number(id: str, ids: np.ndarray) -> np.ndarray:
+    return np.where(ids == id)[0][0]
+
+
+def cosine_similarities(
+    title_embeddings: np.ndarray, abs_embeddings: np.ndarray, row: int
+) -> tuple:
+    return np.sum(title_embeddings[row] * title_embeddings, axis=1), np.sum(
+        abs_embeddings[row] * abs_embeddings, axis=1
+    )
+
+
+def top_hits(
+    title_embeddings: np.ndarray,
+    abs_embeddings: np.ndarray,
+    cat_ohe: np.ndarray,
+    row: int,
+    n: int = 10,
+) -> np.ndarray:
+    assert (
+        title_embeddings.shape[0] == abs_embeddings.shape[0]
+    ), "Not the same number of titles and abstracts"
+
+    """Top n hits that match a row in embeddings, returns their indices
+    and cosine similarities"""
+    cos_similarities = cosine_similarities(title_embeddings, abs_embeddings, row)
```

Comment:
You could generate one vector concatenating all three and use the optimized cosine similarity (or better, just the scalar product) routine from torch. So each row in your embedding would be (title_embedding, abs_embedding, ohe_cat).

Reply:
Looking into that

Reply:
Implemented in semantic_search.py